transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy

```diff
@@ -12,68 +18,85 @@ import org.transmartproject.db.multidimquery.query.InvalidQueryException
+        def job = DetachedCriteria.forClass(AsyncJobCoreDb)
+                .add(Restrictions.idEq(id))
+                .add(Restrictions.eq('userId', user.username))
+                .add(Restrictions.eq('jobType', EXPORT_JOB_TYPE))
+                .getExecutableCriteria(sessionFactory.currentSession)
+                .uniqueResult() as AsyncJobCoreDb
 
-    AsyncJobCoreDb getJobById(Long id) {
-        def job = AsyncJobCoreDb.findById(id)
         if (!job) {
             throw new NoSuchResourceException("Job with id '$id' does not exist.")
         }
         return job
     }
 
+    @Transactional(readOnly = true)
     List<AsyncJobCoreDb> getJobList(User user) {
-        // TODO add time constraints for retrived jobs
-        if(user.admin) {
-            AsyncJobCoreDb.findAllByJobType(jobType)
-        } else {
-            AsyncJobCoreDb.findAllByUserIdAndJobType(user.username, jobType)
-        }
+        DetachedCriteria.forClass(AsyncJobCoreDb)
+                .add(Restrictions.eq('userId', user.username))
+                .add(Restrictions.eq('jobType', EXPORT_JOB_TYPE))
+                .getExecutableCriteria(sessionFactory.currentSession)
+                .list() as List<AsyncJobCoreDb>
     }
 
-    Boolean isJobNameUniqueForUser(String name, User user) {
-        getJobByNameAndUser(name, user) == null
-    }
-
-    AsyncJobCoreDb getJobByNameAndUser(String jobName, User user) {
-        def jobs = AsyncJobCoreDb.findAllByJobNameAndUserId(jobName, user.username)
-        if(jobs.size() > 1) {
-            throw new InvalidQueryException("More than one job with a name '$jobName' found")
+    AsyncJobCoreDb createNewJob(String jobName, User user) {
+        // Check input
+        if (jobName != null) {
+            jobName = jobName.trim()
+            if (jobName.empty) {
+                jobName = null
+            }
         }
-        jobs[0]
-    }
-
-    AsyncJobCoreDb createNewJob(User user, String jobName = null) {
-
-        def jobStatus = JobStatus.CREATED
 
+        // Save job
         def newJob = new AsyncJobCoreDb(lastRunOn: new Date())
-        newJob.jobType = jobType
-        newJob.jobStatus = jobStatus.value
+        newJob.jobType = EXPORT_JOB_TYPE
+        newJob.jobStatus = JobStatus.CREATED.value
         newJob.jobName = jobName
         newJob.userId = user.username
         newJob.save()
 
-        if (!jobName?.trim()) {
+        // Rename job to job id if no job name has been specified
+        if (!jobName) {
             newJob.jobName = newJob.id.toString()
             newJob.save()
         }
 
-        log.debug("Sending ${newJob.jobName} back to the client")
+        // Check for uniqueness of the job name for the user
+        def jobCount = DetachedCriteria.forClass(AsyncJobCoreDb)
+                .add(Restrictions.eq('jobName', newJob.jobName))
+                .add(Restrictions.eq('userId', user.username))
+                .getExecutableCriteria(sessionFactory.currentSession)
```

Comment:
so just to make sure, this will always find the new job that is saved in the session only (not flushed yet) and, if any, some result in the database? There should not be any gorm issues with mixing sessions or sth?

Reply:
Good question. My reasoning here is that this happens within a transaction (because of the `@Transactional` annotation), so within the same transaction it tries to save and check if the resulting number of jobs does not violate uniqueness. If that fails, it rolls back the transaction (reverting the save). For multiple simulatenous requests (with different sessions perhaps), I rely on the database to handle transactions correctly. Hopefully that works as intended...